### PR TITLE
Fix name collision and write additional cpp wrapper unit tests

### PR DESCRIPTION
**Pull request by danaack**
Cpp wrapper tests now additionally include delete, update, equality (partial), range (partial), and all records (partial). 7 unit tests total.

### CODE REVIEW

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 1]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		dict->insert(i, i * 2);
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+
+		dict->deleteRecord(i);
+		int ret_val = dict->get(i);
+
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok != dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 0, dict->last_status.count);
+		PLANCK_UNIT_ASSERT_INT_ARE_NOT_EQUAL(tc, i * 2, ret_val);
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Don't do this assertion - The stuff retruend from `dict->get` is assumed to be not valid here. This should probably fix the Dr. Memory warnings too.

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 2]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+) {
+	for (int i = 0; i < 10; i++) {
+		dict->insert(i, i * 2);
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+
+		dict->deleteRecord(i);
+		int ret_val = dict->get(i);
+
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok != dict->last_status.error);
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Do a stricter check here and assert that `last_status.error` returns `err_item_not_found`.

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 3]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		Dictionary<int, int> *dict
+) {
+	int ret_val;
+
+	dict->insert(-10, 3);
+	PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+	PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+
+	dict->deleteRecord(-10);
+	ret_val = dict->get(-10);
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Similar issue here

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 4]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+
+	dict->deleteRecord(1000);
+	ret_val = dict->get(1000);
+	PLANCK_UNIT_ASSERT_TRUE(tc, err_ok != dict->last_status.error);
+	PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 0, dict->last_status.count);
+	PLANCK_UNIT_ASSERT_INT_ARE_NOT_EQUAL(tc, 99, ret_val);
+}
+
+/**
+@brief	Aggregate test to test insertion on all implementations.
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Fix this comment

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 5]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+	delete dict;
+
+	dict = new OpenAddressFileHash<int, int>(key_type_numeric_signed, sizeof(int), sizeof(int), 50);
+	test_cpp_wrapper_insert_delete(tc, dict);
+	test_cpp_wrapper_insert_delete_edge_cases(tc, dict);
+	delete dict;
+}
+
+/**
+@brief	Tests an insertion and then attempts to delete the record.
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Fix comment

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 6]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		int ret_val = dict->get(i);
+
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, i, ret_val);
+	}
+}
+
+/**
+@brief	Tests insert/delete edge cases such as negative keys and big-digit keys.
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Fix comment

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 7]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+
+	dict->update(1000, 9);
+	ret_val = dict->get(1000);
+	PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+	PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+	PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 9, ret_val);
+}
+
+/**
+@brief	Aggregate test to test insertion on all implementations.
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Fix comment

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 8]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+	delete dict;
+
+	dict = new OpenAddressFileHash<int, int>(key_type_numeric_signed, sizeof(int), sizeof(int), 50);
+	test_cpp_wrapper_insert_update(tc, dict);
+	test_cpp_wrapper_insert_update_edge_cases(tc, dict);
+	delete dict;
+}
+
+void
+test_cpp_wrapper_equality(
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Add an `@brief` comment to this test

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 9]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		dict->insert(i, i*2);
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+	}
+
+	Cursor<int, int> *eq_cursor = dict->equality(eq_key);
+	PLANCK_UNIT_ASSERT_TRUE(tc, eq_cursor->hasNext());
+	eq_cursor->next();
+
+	while(eq_cursor->hasNext()) {
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: This isn't actually the right way to write the Cursor loop - Check out https://github.com/iondbproject/iondb/blob/development/src/examples/iondb_example.ino#L39

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 10]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+
+		dict->insert(i, i*2);
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+	}
+
+	Cursor<int, int> *eq_cursor = dict->equality(eq_key);
+	PLANCK_UNIT_ASSERT_TRUE(tc, eq_cursor->hasNext());
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: This is a little scary to me. Better would be to do an explicit comparison: Use `true == eq_cursor->hasNext()` as the condition instead

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 11]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+
+	while(eq_cursor->hasNext()) {
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, eq_key, eq_cursor->getKey());
+		PLANCK_UNIT_ASSERT_TRUE(tc, eq_cursor->getValue() == record_value);
+		PLANCK_UNIT_ASSERT_TRUE(tc, eq_cursor->hasNext());
+		eq_cursor->next();
+		record_value++;
+	}
+
+	PLANCK_UNIT_ASSERT_TRUE(tc, !eq_cursor->hasNext());
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Same as above, assert against false

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 12]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+
+	/* Check that same number of records are found as were inserted with desired key. */
+	PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, eq_key, record_value);
+	delete eq_cursor;
+}
+
+void
+test_cpp_wrapper_equality_on_all_implementations(
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Add an `@brief` line to this

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 13]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+//	test_cpp_wrapper_equality(tc, dict);
+//	delete dict;
+
+//	dict = new OpenAddressFileHash<int, int>(key_type_numeric_signed, sizeof(int), sizeof(int), 50);
+//	test_cpp_wrapper_equality(tc, dict);
+//	delete dict;
+}
+
+void
+test_cpp_wrapper_range(
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Need `@brief` line

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 14]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+	int max_key = 5;
+
+	for(int i=0; i<10; i++) {
+		dict->insert(i, i);
+		PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 1, dict->last_status.count);
+	}
+
+	Cursor<int, int> *range_cursor = dict->range(min_key, max_key);
+	PLANCK_UNIT_ASSERT_TRUE(tc, range_cursor->hasNext());
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Same

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 15]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+
+	while(range_cursor->hasNext()) {
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, min_key, range_cursor->getKey());
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, range_cursor->getValue(), min_key);
+		PLANCK_UNIT_ASSERT_TRUE(tc, range_cursor->hasNext());
+		range_cursor->next();
+		min_key++;
+	}
+
+	PLANCK_UNIT_ASSERT_TRUE(tc, !range_cursor->hasNext());
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Same

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 16]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+	PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 5, --min_key);
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Should subtract prior to the assert call and **document** why this is necessary. Otherwise it looks spooky.

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 17]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+	delete range_cursor;
+}
+
+void
+test_cpp_wrapper_range_on_all_implementations(
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Give `@brief` line here

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 18]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		planck_unit_test_t *tc
+) {
+	Dictionary<int, int> *dict;
+
+	dict = new BppTree<int, int>(key_type_numeric_signed, sizeof(int), sizeof(int));
+	test_cpp_wrapper_range(tc, dict);
+	delete dict;
+
+//	dict = new SkipList<int, int>(key_type_numeric_signed, sizeof(int), sizeof(int), 7);
+//	test_cpp_wrapper_range(tc, dict);
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Comment on why these tests are commented out

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 19]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+//	delete dict;
+
+//	dict = new OpenAddressFileHash<int, int>(key_type_numeric_signed, sizeof(int), sizeof(int), 50);
+//	test_cpp_wrapper_range(tc, dict);
+//	delete dict;
+}
+
+void
+test_cpp_wrapper_all_records(
+		planck_unit_test_t	*tc,
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Need `@brief`

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 20]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		Dictionary<int, int> *dict
+) {
+	int total_records = 0;
+
+	for(int i=0; i<10; i++) {
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Change up these numbers between tests so that we get a good spread of keys tested.

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 21]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+
+	while(all_rec_cursor->hasNext()) {
+		PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, total_records, all_rec_cursor->getKey());
+		PLANCK_UNIT_ASSERT_TRUE(tc, all_rec_cursor->getValue() == total_records*2);
+		PLANCK_UNIT_ASSERT_TRUE(tc, all_rec_cursor->hasNext());
+		all_rec_cursor->next();
+		total_records++;
+	}
+
+	PLANCK_UNIT_ASSERT_TRUE(tc, !all_rec_cursor->hasNext());
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Same

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 22]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+
+	/* Check that same number of records are found as were inserted with desired key. */
+	PLANCK_UNIT_ASSERT_INT_ARE_EQUAL(tc, 10, total_records);
+	delete all_rec_cursor;
+}
+
+void
+test_cpp_wrapper_all_records_on_all_implementations(
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: `@brief`

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 23]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+//	test_cpp_wrapper_all_records(tc, dict);
+//	delete dict;
+
+//	dict = new OpenAddressFileHash<int, int>(key_type_numeric_signed, sizeof(int), sizeof(int), 50);
+//	test_cpp_wrapper_all_records(tc, dict);
+//	delete dict;
+}
+
+void
+test_cpp_wrapper_open_close(
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: `@brief`

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 24]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		planck_unit_test_t	*tc,
+		Dictionary<int, char*> *dict
+) {
+	ion_status_t status;
+	err_t error;
+	ion_dictionary_id_t gdict_id = dict->dict.instance->id;
+
+	/* Insert test record so we can check data integrity after we close/open */
+	status = dict->insert(66650, GTEST_DATA);
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Lets not use this `GTEST_DATA` define and use our own data

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 25]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+	PLANCK_UNIT_ASSERT_TRUE(tc, err_ok == dict->last_status.error);
+	PLANCK_UNIT_ASSERT_TRUE(tc, 1 == dict->last_status.count);
+
+	int j = memcmp(GTEST_DATA, ret_val, dict->size_v);
+
+	PLANCK_UNIT_ASSERT_TRUE(tc, 0 == j);
+}
+
+void
+test_cpp_wrapper_open_close_on_all_implementations(
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: `@brief`

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 26]
```diff
@@ -112,6 +113,439 @@ test_cpp_wrapper_insert_on_all_implementations(
...
+		planck_unit_test_t *tc
+) {
+	Dictionary<int, char*> *dict;
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Lets keep things consistent and stick with `int` key and `int` values for now

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp [hunk 27]
```diff
@@ -125,6 +559,12 @@ cpp_wrapper_getsuite(
 
 	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_create_and_destroy);
 	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_insert_on_all_implementations);
+	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_delete_on_all_implementations);
+	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_update_on_all_implementations);
+	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_equality_on_all_implementations);
+	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_range_on_all_implementations);
+	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_all_records_on_all_implementations);
+/*	PLANCK_UNIT_ADD_TO_SUITE(suite, test_cpp_wrapper_open_close_on_all_implementations);*/
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.cpp]: Document why this is commented out

File: src/tests/unit/cpp_wrapper/test_cpp_wrapper.h [hunk 28]
```diff
@@ -1,6 +1,8 @@
 #ifndef TEST_CPP_WRAPPER_H_
 #define TEST_CPP_WRAPPER_H_
 
+#define GTEST_DATA (char*) "Test data, please ignore! 123 123 abc abc"
```
Comment (Stickerpants) [inline src/tests/unit/cpp_wrapper/test_cpp_wrapper.h]: Lets remove this define for now

**Comment by Stickerpants** [comment]
LGTM

**Comment by wpenson** [comment]
LGTM
